File: ios/ConstellationSDK/ConstellationSDK/WebView Engine/FormHandler.swift
import Foundation
import WebKit
import OSLog

class FormHandler: NSObject, WKScriptMessageHandler {
    weak var manager: ComponentManager?
    private let continuation = DeferedContinuation<CaseProcessingResult>()

    init(
        manager: ComponentManager
    ) {
        self.manager = manager
    }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard 
            let array = message.body as? [Any],
            let type = array[0] as? String
        else {
            Logger.current().error("Can not decode message")
            return
        }
        switch type {
        case "updateComponent": handleUpdateComponent(array)
        case "addComponent": handleAddComponent(array)
        case "removeComponent": handleRemoveComponent(array)
        case "ready": handleFormReady()
        case "finished": handleFormFinished(array)
        case "cancelled": handleFormCancel()
        case "error": handleFormError(array)
        default:
            Logger.current().error("Unexpected message type: \(type)")
        }
    }

    func processingResult() async -> CaseProcessingResult {
        await continuation.result()
    }

    private func handleUpdateComponent(_ input: [Any]) {
        if
            let cId = input.componentId,
            let props = input[2] as? String {
            manager?.updateComponent(cId, props)
        } else {
            Logger.current().error("Unexpected parameters types in updateComponent")
        }
    }

    private func handleAddComponent(_ input: [Any]) {
        guard let cId = input.componentId, let cType = input[2] as? String else {
            Logger.current().error("Unexpected input for addComponent.")
            return
        }
        manager?.addComponent(cId, type: cType)
    }

    private func handleRemoveComponent(_ input: [Any]) {
        guard let cId = input.componentId  else {
            Logger.current().error("Unexpected input for removeComponent.")
            return
        }
        manager?.removeComponent(cId)
    }

    private func handleFormReady() {
        Logger.current().debug("Form ready.")
    }

    private func handleFormFinished(_ input: [Any]) {
        Logger.current().debug("Form finished.")
        continuation.proceed(.finished(input[1] as? String))
    }

    private func handleFormCancel() {
        Logger.current().debug("Form cancelled.")
        continuation.proceed(.cancelled)
    }

    private func handleFormError(_ input: [Any]) {
        let errorMessage = input[1] as? String ?? "Unexpected."
        Logger.current().debug("Form encountered an error: \(errorMessage)")
        continuation.proceed(.error(errorMessage))
    }
}

extension Array where Element == Any {
    fileprivate var componentId: String? {
        if let intId = self[1] as? Int {
            "\(intId)"
        } else if let stringId = self[1] as? String {
            stringId
        } else {
            nil
        }
    }
}

File: core/src/iosMain/kotlin/com/pega/constellation/sdk/kmp/core/engine/FormHandler.kt
package com.pega.constellation.sdk.kmp.core.engine

import com.pega.constellation.sdk.kmp.core.EngineEvent
import com.pega.constellation.sdk.kmp.core.EngineEventHandler
import com.pega.constellation.sdk.kmp.core.Log
import com.pega.constellation.sdk.kmp.core.api.ComponentId
import com.pega.constellation.sdk.kmp.core.api.ComponentManager
import com.pega.constellation.sdk.kmp.core.api.ComponentType
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import platform.WebKit.*
import platform.darwin.NSObject

private const val TAG = "FormHandler"
class FormHandler(
    private val eventHandler: EngineEventHandler,
    private val componentManager: ComponentManager
) : NSObject(), WKScriptMessageHandlerProtocol {

    private val passthroughSubject = MutableSharedFlow<ComponentEvent>(extraBufferCapacity = 64)
    val eventStream: SharedFlow<ComponentEvent> = passthroughSubject

    override fun userContentController(
        userContentController: WKUserContentController,
        didReceiveScriptMessage: WKScriptMessage
    ) {
        val array = didReceiveScriptMessage.body as? List<Any?> ?: run {
            Log.w(TAG, "Cannot decode message")
            return
        }
        val type = array.getOrNull(0) as? String ?: run {
            Log.w(TAG, "Cannot decode message type")
            return
        }
        when (type) {
            "updateComponent" -> handleUpdateComponent(array)
            "addComponent" -> handleAddComponent(array)
            "removeComponent" -> handleRemoveComponent(array)
            "ready" -> eventHandler.handle(EngineEvent.Ready)
            "finished" -> eventHandler.handle(EngineEvent.Finished(array.getOrNull(1) as? String))
            "cancelled" -> eventHandler.handle(EngineEvent.Cancelled)
            "error" -> eventHandler.handle(EngineEvent.Error(array.getOrNull(1) as? String))
            else -> Log.w(TAG, "Unexpected message type: $type")
        }
    }

    fun handleLoading() {
        eventHandler.handle(EngineEvent.Loading)
    }

    private fun handleUpdateComponent(input: List<Any?>) {

        val cId = input.componentId
        val props = input.getOrNull(2) as? String
        if (cId != null && props != null) {
            val propsJson = Json.parseToJsonElement(props).jsonObject
            componentManager.updateComponent(ComponentId(cId), propsJson)
        } else {
            Log.w(TAG, "Unexpected parameters types in updateComponent")
        }
    }

    private fun handleAddComponent(input: List<Any?>) {
        val cId = input.componentId
        val cType = input.getOrNull(2) as? String
        if (cId == null || cType == null) {
            Log.w(TAG, "Unexpected input for addComponent.")
            return
        }

        val context = WKWebViewEngineComponentContext(
            ComponentId(cId),
            ComponentType(cType),
            componentManager) { eventContent ->
                passthroughSubject.tryEmit(ComponentEvent(cId, eventContent))
        }
        componentManager.addComponent(context)
    }

    private fun handleRemoveComponent(input: List<Any?>) {
        val cId = input.componentId
        if (cId == null) {
            Log.w(TAG, "Unexpected input for removeComponent.")
            return
        }
        componentManager.removeComponent(ComponentId(cId))
    }
}

val List<Any?>.componentId: Int?
    get() = when (val value = this.getOrNull(1)) {
        is Int -> value
        is String -> value.toIntOrNull()
        else -> null
    }

